%\VignetteIndexEntry{The GenomeGraphs users guide}
%\VignetteDepends{GenomeGraphs}
%\VignetteKeywords{Visualization}
%\VignettePackage{GenomeGraphs}
\documentclass[11pt]{article}
\usepackage{hyperref}
\usepackage{url}
\usepackage[authoryear,round]{natbib}
\bibliographystyle{plainnat}

\newcommand{\scscst}{\scriptscriptstyle}
\newcommand{\scst}{\scriptstyle}

\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textit{#1}}}


\author{Steffen Durinck\footnote{steffen@stat.berkeley.edu} and James
  Bullard\footnote{bullard@stat.berkeley.edu}}
\usepackage{Sweave}
\begin{document}
\title{The GenomeGraphs user's guide}

\maketitle

\tableofcontents
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Introduction}

Genomic data analyses can benifit from integrated visualization of the
genomic information.  The GenomeGraphs package uses the biomaRt
package to do live queries to Ensembl and translates
e.g. gene/transcript structures to viewports of the grid graphics
package, resulting in genomic information plotted together with your
data.  Possible genomics datasets that can be plotted are: Array CGH
data, gene expression data and sequencing data.

\begin{Schunk}
\begin{Sinput}
> library(GenomeGraphs)
\end{Sinput}
\end{Schunk}


\section{Creating a Ensembl annotation graphic}

To create an Ensembl annotation graphic, you need to decide what you
want to plot.  Genes and transcripts can be plotted individually using
the \Robject{Gene} and \Robject{Transcript} objects respectively.  Or
one can plot a gene region the forward strand or reverse strand only
or both.  In this section we will cover these different graphics.

\subsection{Plotting a Gene}

If one wants to plot annotation information from Ensembl then you need
to connect to the Ensembl BioMart database using the useMart function
of the biomaRt package.

\begin{Schunk}
\begin{Sinput}
> mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
\end{Sinput}
\end{Schunk}

Next we can retrieve the gene structure of the gene of interest.

\begin{Schunk}
\begin{Sinput}
> gene <- makeGene(id = "ENSG00000095203", 
+     type = "ensembl_gene_id", biomart = mart)
> gdPlot(gene)
\end{Sinput}
\end{Schunk}
\includegraphics{GenomeGraphs-005}

\subsection{Adding alternative transcripts}

To add alternative transcripts you first have to create a \Robject{Transcript} object.
Note that the order of the objects in the list determines the order in the plot.
\begin{Schunk}
\begin{Sinput}
> transcript <- makeTranscript(id = "ENSG00000095203", 
+     type = "ensembl_gene_id", biomart = mart)
> gdPlot(list(gene, transcript))
\end{Sinput}
\end{Schunk}
\includegraphics{GenomeGraphs-006}

\subsection{Plotting a gene region}

If you're interested in not just plotting one gene but a whole gene
region the you should create a \Robject{GeneRegion} object.  Note that
a \Robject{GeneRegion} object is strand specific.  In the example
below we will retrieve the genes on the forward (+) strand only and
add a genomic axis as well to give us the base positions.

\begin{Schunk}
\begin{Sinput}
> plusStrand <- makeGeneRegion(chromosome = 17, 
+     start = 30450000, end = 30550000, 
+     strand = "+", biomart = mart)